
Use libdir from configuration.

--- httpd-2.4.18/support/apxs.in.apxs
+++ httpd-2.4.18/support/apxs.in
@@ -35,6 +35,7 @@ my $CFG_PREFIX     = $prefix;
 my $exec_prefix    = get_vars("exec_prefix");
 my $datadir        = get_vars("datadir");
 my $localstatedir  = get_vars("localstatedir");
+my $libdir         = get_vars("libdir");
 my $CFG_TARGET     = get_vars("progname");
 my $CFG_SYSCONFDIR = get_vars("sysconfdir");
 my $CFG_CFLAGS     = join ' ', map { get_vars($_) }
@@ -275,7 +276,7 @@ if ($opt_g) {
     $data =~ s|%NAME%|$name|sg;
     $data =~ s|%TARGET%|$CFG_TARGET|sg;
     $data =~ s|%PREFIX%|$prefix|sg;
-    $data =~ s|%INSTALLBUILDDIR%|$installbuilddir|sg;
+    $data =~ s|%LIBDIR%|$libdir|sg;
 
     my ($mkf, $mods, $src) = ($data =~ m|^(.+)-=#=-\n(.+)-=#=-\n(.+)|s);
 
@@ -453,11 +454,11 @@ if ($opt_c) {
     my $ldflags = "$CFG_LDFLAGS";
     if ($opt_p == 1) {
         
-        my $apr_libs=`$apr_config --cflags --ldflags --link-libtool --libs`;
+        my $apr_libs=`$apr_config --cflags --ldflags --link-libtool`;
         chomp($apr_libs);
         my $apu_libs="";
         if ($apr_major_version < 2) {
-            $apu_libs=`$apu_config --ldflags --link-libtool --libs`;
+            $apu_libs=`$apu_config --ldflags --link-libtool`;
             chomp($apu_libs);
         }
         
@@ -672,8 +673,8 @@ __DATA__
 
 builddir=.
 top_srcdir=%PREFIX%
-top_builddir=%PREFIX%
-include %INSTALLBUILDDIR%/special.mk
+top_builddir=%LIBDIR%/httpd
+include %LIBDIR%/httpd/build/special.mk
 
 #   the used tools
 APACHECTL=apachectl
